File: src/codex_autorunner/web_frontend/src/lib/api/readModelContracts.ts
export const READ_MODEL_CONTRACT_VERSION = 'web-read-models.v1' as const;

export type ReadModelContractVersion = typeof READ_MODEL_CONTRACT_VERSION;

export type ProjectionCursor = {
  value: string;
  sequence: number;
  source: string;
  issuedAt: string;
};

export type ProjectionRevision = {
  value: string;
  sourceKind: string;
  sourceId: string;
  updatedAt: string;
};

export type RepairPolicy = {
  snapshotRoute: string;
  cursorQueryParam: 'after';
  gapEventType: 'projection.cursor_gap';
  behavior: 'repair_snapshot_required';
};

export type PageWindow = {
  limit: number;
  nextCursor?: string | null;
  previousCursor?: string | null;
  totalEstimate?: number | null;
  totalIsExact: boolean;
};

export type ReadModelOperation = 'upsert' | 'patch' | 'delete' | 'reorder' | 'invalidate' | 'reset';

export type ReadModelEventEnvelope<TEvent extends string = string, TEntityKind extends string = string> = {
  contractVersion: ReadModelContractVersion;
  eventType: TEvent;
  cursor: ProjectionCursor;
  entityKind: TEntityKind;
  entityId: string;
  operation: ReadModelOperation;
  generatedAt: string;
  sourceRevision?: ProjectionRevision | null;
};

export type ChatIndexRow = {
  chatId: string;
  surface: 'pma' | 'file_chat' | 'telegram' | 'discord' | 'app_server' | 'other';
  title: string;
  status: 'waiting' | 'running' | 'idle' | 'archived' | 'failed';
  unreadCount: number;
  lastActivityAt?: string | null;
  repoId?: string | null;
  worktreeId?: string | null;
  ticketId?: string | null;
  runId?: string | null;
  agent?: string | null;
  model?: string | null;
  groupId?: string | null;
};

export type ChatIndexGroup = {
  groupId: string;
  kind: 'ticket_run' | 'surface' | 'repo' | 'worktree';
  label: string;
  childCount: number;
  expandedChildWindow?: PageWindow | null;
};

export type ChatIndexCounters = {
  total: number;
  waiting: number;
  running: number;
  unread: number;
  archived: number;
};

export type ChatIndexSnapshot = {
  contractVersion: ReadModelContractVersion;
  kind: 'chat.index.snapshot';
  cursor: ProjectionCursor;
  window: PageWindow;
  filter: 'all' | 'waiting' | 'active' | 'unread' | 'archived' | 'ticket_runs' | 'external';
  query?: string | null;
  rows: ChatIndexRow[];
  groups: ChatIndexGroup[];
  counters: ChatIndexCounters;
  repair: RepairPolicy;
};

export type ChatIndexPatch = {
  rows: ChatIndexRow[];
  groups: ChatIndexGroup[];
  removedRowIds: string[];
  removedGroupIds: string[];
  order?: string[] | null;
  counters?: ChatIndexCounters | null;
};

export type ChatIndexPatchEvent = {
  envelope: ReadModelEventEnvelope<'chat.index.patch', 'chat'>;
  patch: ChatIndexPatch;
};

export type ChatTimelineItem = {
  itemId: string;
  kind: 'user_message' | 'assistant_message' | 'tool_event' | 'progress' | 'artifact' | 'system';
  role?: 'user' | 'assistant' | 'tool' | 'system' | null;
  createdAt: string;
  text?: string | null;
  artifactIds: string[];
  clientMessageId?: string | null;
  backendMessageId?: string | null;
};

export type ChatQueueSummary = {
  depth: number;
  activeTurnId?: string | null;
  queuedTurnIds: string[];
};

export type ChatArtifactSummary = {
  artifactId: string;
  name: string;
  kind: string;
  href?: string | null;
  updatedAt?: string | null;
};

export type ChatThreadProjection = {
  chatId: string;
  surface: string;
  title: string;
  status: 'waiting' | 'running' | 'idle' | 'archived' | 'failed';
  repoId?: string | null;
  worktreeId?: string | null;
  ticketId?: string | null;
  runId?: string | null;
  agent?: string | null;
  model?: string | null;
  archived: boolean;
};

export type ChatDetailSnapshot = {
  contractVersion: ReadModelContractVersion;
  kind: 'chat.detail.snapshot';
  cursor: ProjectionCursor;
  thread: ChatThreadProjection;
  timelineWindow: PageWindow;
  timeline: ChatTimelineItem[];
  queue: ChatQueueSummary;
  artifacts: ChatArtifactSummary[];
  repair: RepairPolicy;
};

export type ChatDetailPatch = {
  thread?: ChatThreadProjection | null;
  appendedTimeline: ChatTimelineItem[];
  patchedTimeline: ChatTimelineItem[];
  removedTimelineIds: string[];
  queue?: ChatQueueSummary | null;
  artifacts: ChatArtifactSummary[];
};

export type ChatDetailPatchEvent = {
  envelope: ReadModelEventEnvelope<'chat.detail.patch', 'chat'>;
  patch: ChatDetailPatch;
};

export type RepoTopology = {
  repoId: string;
  label: string;
  path: string;
  archived: boolean;
  destinationId?: string | null;
  childWorktreeIds: string[];
};

export type WorktreeTopology = {
  worktreeId: string;
  repoId: string;
  label: string;
  path: string;
  branch?: string | null;
  archived: boolean;
  destinationId?: string | null;
};

export type RepoWorktreeTopologySnapshot = {
  contractVersion: ReadModelContractVersion;
  kind: 'repo_worktree.topology.snapshot';
  cursor: ProjectionCursor;
  window: PageWindow;
  repos: RepoTopology[];
  worktrees: WorktreeTopology[];
  repair: RepairPolicy;
};

export type RuntimeProjection = {
  entityKind: 'repo' | 'worktree';
  entityId: string;
  gitDirty?: boolean | null;
  gitAhead?: number | null;
  gitBehind?: number | null;
  activeRunId?: string | null;
  activeRunStatus?: string | null;
  waitingTicketCount: number;
  runningTicketCount: number;
  chatCount: number;
  cleanupBlockers: string[];
  updatedAt?: string | null;
};

export type RepoWorktreeRuntimeSnapshot = {
  contractVersion: ReadModelContractVersion;
  kind: 'repo_worktree.runtime.snapshot';
  cursor: ProjectionCursor;
  window: PageWindow;
  runtime: RuntimeProjection[];
  repair: RepairPolicy;
};

export type RepoWorktreePatch = {
  topologyRepos: RepoTopology[];
  topologyWorktrees: WorktreeTopology[];
  runtime: RuntimeProjection[];
  removedRepoIds: string[];
  removedWorktreeIds: string[];
  order?: string[] | null;
};

export type RepoWorktreePatchEvent = {
  envelope: ReadModelEventEnvelope<'repo.topology.patch' | 'repo.runtime.patch' | 'worktree.topology.patch' | 'worktree.runtime.patch', 'repo' | 'worktree'>;
  patch: RepoWorktreePatch;
};

export type RepoWorktreeDetailSnapshot = {
  contractVersion: ReadModelContractVersion;
  kind: 'repo_worktree.detail.snapshot';
  cursor: ProjectionCursor;
  ownerKind: 'repo' | 'worktree';
  ownerId: string;
  identity: Record<string, unknown>;
  parentLinks: Record<string, unknown>;
  topology: Record<string, unknown>;
  runtime: Record<string, unknown>;
  scopedTickets: Record<string, unknown>[];
  scopedRuns: Record<string, unknown>[];
  scopedChats: Record<string, unknown>[];
  contextspaceSummary: Record<string, unknown>[];
  currentArtifacts: Record<string, unknown>[];
  ticketWindow: PageWindow;
  runWindow: PageWindow;
  chatWindow: PageWindow;
  artifactWindow: PageWindow;
  repair: RepairPolicy;
};

export type TicketProjection = {
  ticketId: string;
  routeId: string;
  title: string;
  status: 'queued' | 'waiting' | 'running' | 'blocked' | 'done' | 'failed' | 'invalid';
  ownerKind: 'repo' | 'worktree';
  ownerId: string;
  agent?: string | null;
  model?: string | null;
  done: boolean;
  updatedAt?: string | null;
};

export type TicketQueueSibling = {
  ticketId: string;
  routeId: string;
  title: string;
  status: string;
  previousTicketId?: string | null;
  nextTicketId?: string | null;
};

export type RunProjection = {
  runId: string;
  status: string;
  startedAt?: string | null;
  finishedAt?: string | null;
  workerActivity?: string | null;
};

export type TicketDetailSnapshot = {
  contractVersion: ReadModelContractVersion;
  kind: 'ticket.detail.snapshot';
  cursor: ProjectionCursor;
  ticket: TicketProjection;
  siblings: TicketQueueSibling[];
  linkedRun?: RunProjection | null;
  linkedChats: ChatIndexRow[];
  artifacts: ChatArtifactSummary[];
  dispatchWindow: PageWindow;
  dispatches: Record<string, unknown>[];
  repair: RepairPolicy;
};

export type TicketDetailPatch = {
  ticket?: TicketProjection | null;
  siblings: TicketQueueSibling[];
  linkedRun?: RunProjection | null;
  linkedChats: ChatIndexRow[];
  artifacts: ChatArtifactSummary[];
  dispatches: Record<string, unknown>[];
};

export type TicketDetailPatchEvent = {
  envelope: ReadModelEventEnvelope<'ticket.detail.patch', 'ticket'>;
  patch: TicketDetailPatch;
};

export type ReadModelSnapshot =
  | ChatIndexSnapshot
  | ChatDetailSnapshot
  | RepoWorktreeTopologySnapshot
  | RepoWorktreeRuntimeSnapshot
  | RepoWorktreeDetailSnapshot
  | TicketDetailSnapshot;

export type ReadModelPatchEvent =
  | ChatIndexPatchEvent
  | ChatDetailPatchEvent
  | RepoWorktreePatchEvent
  | TicketDetailPatchEvent;

export function mapReadModelContract<T extends ReadModelSnapshot | ReadModelPatchEvent>(payload: unknown): T {
  const record = asRecord(payload);
  const version = record.contractVersion ?? asRecord(record.envelope).contractVersion;
  if (version !== READ_MODEL_CONTRACT_VERSION) {
    throw new Error(`Unsupported read model contract version: ${String(version)}`);
  }
  return JSON.parse(JSON.stringify(record)) as T;
}

export function readModelRepairRequired(event: ReadModelPatchEvent): boolean {
  const eventType: string = event.envelope.eventType;
  return eventType === 'projection.cursor_gap' || event.envelope.operation === 'reset';
}

function asRecord(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error('Expected read model contract object.');
  }
  return value as Record<string, unknown>;
}

File: src/codex_autorunner/surfaces/web/read_model_contracts.py
"""Versioned web UI read-model contracts.

These models define the target screen-shaped payloads for the responsive web UI
projection layer. Route handlers are expected to serialize them with
``dump_read_model_contract`` so JSON field names match the TypeScript contracts.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Literal, Optional, TypeVar

from pydantic import BaseModel, ConfigDict, Field

READ_MODEL_CONTRACT_VERSION: Literal["web-read-models.v1"] = "web-read-models.v1"

__all__ = [
    "READ_MODEL_CONTRACT_VERSION",
    "ChatArtifactSummary",
    "ChatDetailPatch",
    "ChatDetailPatchEvent",
    "ChatDetailSnapshot",
    "ChatIndexCounters",
    "ChatIndexGroup",
    "ChatIndexPatch",
    "ChatIndexPatchEvent",
    "ChatIndexRow",
    "ChatIndexSnapshot",
    "ChatQueueSummary",
    "ChatThreadProjection",
    "ChatTimelineItem",
    "PageWindow",
    "ProjectionCursor",
    "ProjectionRevision",
    "ReadModelContract",
    "ReadModelEventEnvelope",
    "RepairPolicy",
    "RepoTopology",
    "RepoWorktreeDetailSnapshot",
    "RepoWorktreePatch",
    "RepoWorktreePatchEvent",
    "RepoWorktreeRuntimeSnapshot",
    "RepoWorktreeTopologySnapshot",
    "RunProjection",
    "RuntimeProjection",
    "TicketDetailPatch",
    "TicketDetailPatchEvent",
    "TicketDetailSnapshot",
    "TicketProjection",
    "TicketQueueSibling",
    "WorktreeTopology",
    "dump_read_model_contract",
    "load_read_model_contract",
    "read_model_now",
]


def _to_camel(value: str) -> str:
    parts = value.split("_")
    return parts[0] + "".join(part[:1].upper() + part[1:] for part in parts[1:])


class ReadModelContract(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
        populate_by_name=True,
        alias_generator=_to_camel,
    )


class ProjectionCursor(ReadModelContract):
    value: str
    sequence: int = Field(ge=0)
    source: str
    issued_at: datetime


class ProjectionRevision(ReadModelContract):
    value: str
    source_kind: str
    source_id: str
    updated_at: datetime


class RepairPolicy(ReadModelContract):
    snapshot_route: str
    cursor_query_param: str = "after"
    gap_event_type: Literal["projection.cursor_gap"] = "projection.cursor_gap"
    behavior: Literal["repair_snapshot_required"] = "repair_snapshot_required"


class PageWindow(ReadModelContract):
    limit: int = Field(ge=1, le=500)
    next_cursor: Optional[str] = None
    previous_cursor: Optional[str] = None
    total_estimate: Optional[int] = Field(default=None, ge=0)
    total_is_exact: bool = False


class ReadModelEventEnvelope(ReadModelContract):
    contract_version: Literal["web-read-models.v1"] = READ_MODEL_CONTRACT_VERSION
    event_type: str
    cursor: ProjectionCursor
    entity_kind: str
    entity_id: str
    operation: Literal["upsert", "patch", "delete", "reorder", "invalidate", "reset"]
    generated_at: datetime
    source_revision: Optional[ProjectionRevision] = None


class ChatIndexRow(ReadModelContract):
    chat_id: str
    surface: Literal["pma", "file_chat", "telegram", "discord", "app_server", "other"]
    title: str
    status: Literal["waiting", "running", "idle", "archived", "failed"]
    unread_count: int = Field(ge=0)
    last_activity_at: Optional[datetime] = None
    repo_id: Optional[str] = None
    worktree_id: Optional[str] = None
    ticket_id: Optional[str] = None
    run_id: Optional[str] = None
    agent: Optional[str] = None
    model: Optional[str] = None
    group_id: Optional[str] = None


class ChatIndexGroup(ReadModelContract):
    group_id: str
    kind: Literal["ticket_run", "surface", "repo", "worktree"]
    label: str
    child_count: int = Field(ge=0)
    expanded_child_window: Optional[PageWindow] = None


class ChatIndexCounters(ReadModelContract):
    total: int = Field(ge=0)
    waiting: int = Field(ge=0)
    running: int = Field(ge=0)
    unread: int = Field(ge=0)
    archived: int = Field(ge=0)


class ChatIndexSnapshot(ReadModelContract):
    contract_version: Literal["web-read-models.v1"] = READ_MODEL_CONTRACT_VERSION
    kind: Literal["chat.index.snapshot"] = "chat.index.snapshot"
    cursor: ProjectionCursor
    window: PageWindow
    filter: Literal[
        "all", "waiting", "active", "unread", "archived", "ticket_runs", "external"
    ]
    query: Optional[str] = None
    rows: list[ChatIndexRow]
    groups: list[ChatIndexGroup] = Field(default_factory=list)
    counters: ChatIndexCounters
    repair: RepairPolicy


class ChatIndexPatch(ReadModelContract):
    rows: list[ChatIndexRow] = Field(default_factory=list)
    groups: list[ChatIndexGroup] = Field(default_factory=list)
    removed_row_ids: list[str] = Field(default_factory=list)
    removed_group_ids: list[str] = Field(default_factory=list)
    order: Optional[list[str]] = None
    counters: Optional[ChatIndexCounters] = None


class ChatIndexPatchEvent(ReadModelContract):
    envelope: ReadModelEventEnvelope
    patch: ChatIndexPatch


class ChatTimelineItem(ReadModelContract):
    item_id: str
    kind: Literal[
        "user_message",
        "assistant_message",
        "tool_event",
        "progress",
        "artifact",
        "system",
    ]
    role: Optional[Literal["user", "assistant", "tool", "system"]] = None
    created_at: datetime
    text: Optional[str] = None
    artifact_ids: list[str] = Field(default_factory=list)
    client_message_id: Optional[str] = None
    backend_message_id: Optional[str] = None


class ChatQueueSummary(ReadModelContract):
    depth: int = Field(ge=0)
    active_turn_id: Optional[str] = None
    queued_turn_ids: list[str] = Field(default_factory=list)


class ChatArtifactSummary(ReadModelContract):
    artifact_id: str
    name: str
    kind: str
    href: Optional[str] = None
    updated_at: Optional[datetime] = None


class ChatThreadProjection(ReadModelContract):
    chat_id: str
    surface: str
    title: str
    status: Literal["waiting", "running", "idle", "archived", "failed"]
    repo_id: Optional[str] = None
    worktree_id: Optional[str] = None
    ticket_id: Optional[str] = None
    run_id: Optional[str] = None
    agent: Optional[str] = None
    model: Optional[str] = None
    archived: bool = False


class ChatDetailSnapshot(ReadModelContract):
    contract_version: Literal["web-read-models.v1"] = READ_MODEL_CONTRACT_VERSION
    kind: Literal["chat.detail.snapshot"] = "chat.detail.snapshot"
    cursor: ProjectionCursor
    thread: ChatThreadProjection
    timeline_window: PageWindow
    timeline: list[ChatTimelineItem]
    queue: ChatQueueSummary
    artifacts: list[ChatArtifactSummary] = Field(default_factory=list)
    repair: RepairPolicy


class ChatDetailPatch(ReadModelContract):
    thread: Optional[ChatThreadProjection] = None
    appended_timeline: list[ChatTimelineItem] = Field(default_factory=list)
    patched_timeline: list[ChatTimelineItem] = Field(default_factory=list)
    removed_timeline_ids: list[str] = Field(default_factory=list)
    queue: Optional[ChatQueueSummary] = None
    artifacts: list[ChatArtifactSummary] = Field(default_factory=list)


class ChatDetailPatchEvent(ReadModelContract):
    envelope: ReadModelEventEnvelope
    patch: ChatDetailPatch


class RepoTopology(ReadModelContract):
    repo_id: str
    label: str
    path: str
    archived: bool = False
    destination_id: Optional[str] = None
    child_worktree_ids: list[str] = Field(default_factory=list)


class WorktreeTopology(ReadModelContract):
    worktree_id: str
    repo_id: str
    label: str
    path: str
    branch: Optional[str] = None
    archived: bool = False
    destination_id: Optional[str] = None


class RepoWorktreeTopologySnapshot(ReadModelContract):
    contract_version: Literal["web-read-models.v1"] = READ_MODEL_CONTRACT_VERSION
    kind: Literal["repo_worktree.topology.snapshot"] = "repo_worktree.topology.snapshot"
    cursor: ProjectionCursor
    window: PageWindow
    repos: list[RepoTopology]
    worktrees: list[WorktreeTopology]
    repair: RepairPolicy


class RuntimeProjection(ReadModelContract):
    entity_kind: Literal["repo", "worktree"]
    entity_id: str
    git_dirty: Optional[bool] = None
    git_ahead: Optional[int] = Field(default=None, ge=0)
    git_behind: Optional[int] = Field(default=None, ge=0)
    active_run_id: Optional[str] = None
    active_run_status: Optional[str] = None
    waiting_ticket_count: int = Field(default=0, ge=0)
    running_ticket_count: int = Field(default=0, ge=0)
    chat_count: int = Field(default=0, ge=0)
    cleanup_blockers: list[str] = Field(default_factory=list)
    updated_at: Optional[datetime] = None


class RepoWorktreeRuntimeSnapshot(ReadModelContract):
    contract_version: Literal["web-read-models.v1"] = READ_MODEL_CONTRACT_VERSION
    kind: Literal["repo_worktree.runtime.snapshot"] = "repo_worktree.runtime.snapshot"
    cursor: ProjectionCursor
    window: PageWindow
    runtime: list[RuntimeProjection]
    repair: RepairPolicy


class RepoWorktreePatch(ReadModelContract):
    topology_repos: list[RepoTopology] = Field(default_factory=list)
    topology_worktrees: list[WorktreeTopology] = Field(default_factory=list)
    runtime: list[RuntimeProjection] = Field(default_factory=list)
    removed_repo_ids: list[str] = Field(default_factory=list)
    removed_worktree_ids: list[str] = Field(default_factory=list)
    order: Optional[list[str]] = None


class RepoWorktreePatchEvent(ReadModelContract):
    envelope: ReadModelEventEnvelope
    patch: RepoWorktreePatch


class RepoWorktreeDetailSnapshot(ReadModelContract):
    contract_version: Literal["web-read-models.v1"] = READ_MODEL_CONTRACT_VERSION
    kind: Literal["repo_worktree.detail.snapshot"] = "repo_worktree.detail.snapshot"
    cursor: ProjectionCursor
    owner_kind: Literal["repo", "worktree"]
    owner_id: str
    identity: dict[str, Any]
    parent_links: dict[str, Any] = Field(default_factory=dict)
    topology: dict[str, Any] = Field(default_factory=dict)
    runtime: dict[str, Any] = Field(default_factory=dict)
    scoped_tickets: list[dict[str, Any]] = Field(default_factory=list)
    scoped_runs: list[dict[str, Any]] = Field(default_factory=list)
    scoped_chats: list[dict[str, Any]] = Field(default_factory=list)
    contextspace_summary: list[dict[str, Any]] = Field(default_factory=list)
    current_artifacts: list[dict[str, Any]] = Field(default_factory=list)
    ticket_window: PageWindow
    run_window: PageWindow
    chat_window: PageWindow
    artifact_window: PageWindow
    repair: RepairPolicy


class TicketProjection(ReadModelContract):
    ticket_id: str
    route_id: str
    title: str
    status: Literal[
        "queued", "waiting", "running", "blocked", "done", "failed", "invalid"
    ]
    owner_kind: Literal["repo", "worktree"]
    owner_id: str
    agent: Optional[str] = None
    model: Optional[str] = None
    done: bool = False
    updated_at: Optional[datetime] = None


class TicketQueueSibling(ReadModelContract):
    ticket_id: str
    route_id: str
    title: str
    status: str
    previous_ticket_id: Optional[str] = None
    next_ticket_id: Optional[str] = None


class RunProjection(ReadModelContract):
    run_id: str
    status: str
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    worker_activity: Optional[str] = None


class TicketDetailSnapshot(ReadModelContract):
    contract_version: Literal["web-read-models.v1"] = READ_MODEL_CONTRACT_VERSION
    kind: Literal["ticket.detail.snapshot"] = "ticket.detail.snapshot"
    cursor: ProjectionCursor
    ticket: TicketProjection
    siblings: list[TicketQueueSibling] = Field(default_factory=list)
    linked_run: Optional[RunProjection] = None
    linked_chats: list[ChatIndexRow] = Field(default_factory=list)
    artifacts: list[ChatArtifactSummary] = Field(default_factory=list)
    dispatch_window: PageWindow
    dispatches: list[dict[str, Any]] = Field(default_factory=list)
    repair: RepairPolicy


class TicketDetailPatch(ReadModelContract):
    ticket: Optional[TicketProjection] = None
    siblings: list[TicketQueueSibling] = Field(default_factory=list)
    linked_run: Optional[RunProjection] = None
    linked_chats: list[ChatIndexRow] = Field(default_factory=list)
    artifacts: list[ChatArtifactSummary] = Field(default_factory=list)
    dispatches: list[dict[str, Any]] = Field(default_factory=list)


class TicketDetailPatchEvent(ReadModelContract):
    envelope: ReadModelEventEnvelope
    patch: TicketDetailPatch


ReadModel = TypeVar("ReadModel", bound=ReadModelContract)


def read_model_now() -> datetime:
    return datetime.now(timezone.utc)


def dump_read_model_contract(model: ReadModelContract) -> dict[str, Any]:
    return model.model_dump(mode="json", by_alias=True, exclude_none=True)


def load_read_model_contract(
    model_type: type[ReadModel], payload: dict[str, Any]
) -> ReadModel:
    return model_type.model_validate(payload)
